### PR TITLE
Imitation learning support 

### PR DESCRIPTION
**Pull request by Ivan-267**
### Motivation: 
Supporting pre-training with GAIL could be helpful in some environments where it might be difficult to define a good dense reward function that results in the desired behavior. One use case could be to do some pre-training with GAIL followed by standard PPO RL training with a sparse reward function. Of course, making quality recorded demos and setting more hyperparameters are added challenges. The intention of this PR is only to add basic support.

### TODO:
- [x] Onnx export (as the single obs space wrapper is used, it needs a slightly different onnx export script, probably similar to the cleanrl export script)
- [x] Testing (basic)
- [x] Basic guide on how to use

### Description:
Together with demo recorder (https://github.com/edbeeching/godot_rl_agents_plugin/pull/35), adds basic support and example script for training or pre-training using imitation learning.

Uses the imitation library, which is compatible with SB3: https://imitation.readthedocs.io/en/latest/
> Imitation provides clean implementations of imitation and reward learning algorithms, under a unified and user-friendly API. Currently, we have implementations of Behavioral Cloning, [DAgger](https://arxiv.org/pdf/1011.0686.pdf) (with synthetic examples), density-based reward modeling, [Maximum Causal Entropy Inverse Reinforcement Learning](https://www.cs.cmu.edu/~bziebart/publications/maximum-causal-entropy.pdf), [Adversarial Inverse Reinforcement Learning](https://arxiv.org/abs/1710.11248), [Generative Adversarial Imitation Learning](https://arxiv.org/abs/1606.03476), and [Deep RL from Human Preferences](https://arxiv.org/abs/1706.03741).

For now, the script uses GAIL, although BC and AIRL will also work with small changes to the script. Some of the other algorithms may need changes to the env code or different data formats than the current demo recorder makes.

### CODE REVIEW

**Comment by Ivan-267** [comment]
I have added a preliminary tutorial here:
https://github.com/edbeeching/godot_rl_agents/blob/imitation_learning_experimental/docs/IMITATION_LEARNING.md

As there are no example envs built specifically for imitation learning yet (all are designed to work with the reward function), it shows how to modify an existing env to record demos and train an agent from the demos using GAIL. Even though it doesn't improve performance in that specific case, it explains how to setup an env for demo recording and the rest of the process which can be applied to any existing or new / custom env.

**Comment by edbeeching** [comment]
hey @Ivan-267 , is this good to review now?

**Comment by Ivan-267** [comment]
> hey @Ivan-267 thanks for adding this! Is this good to review now?

Yes, it should be.

File: docs/IMITATION_LEARNING.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+		robot.requested_movement = Vector3(
+			int(Input.is_action_pressed("ui_down")) - int(Input.is_action_pressed("ui_up")), 
+			0.0, 
+			int(Input.is_action_pressed("ui_left")) - int(Input.is_action_pressed("ui_right"))
+		).limit_length(1.0)
+```
+
+The way this works is that if we are running training or inference with a RL agent, the `set_action` method will be
+called
+with action values provided. However, during demo recording, `set_acation` will be called without any action provided,
```
Comment (edbeeching) [inline docs/IMITATION_LEARNING.md:115]: typo `set_acation`

File: docs/IMITATION_LEARNING.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+Here's a highly sped-up video of recording 18 episodes:
+
+https://github.com/edbeeching/godot_rl_agents/assets/61947090/7bdc19ba-6e88-431d-b87b-7ec3e0ce1a7c
+
+> [!NOTE]
+> I found it difficult to control the robot with action repeat 10, and I removed a few episodes where the robot hit
+> an enemy robot during recording so that they don't end up in the recorded demos file. I would recommend setting action
+> repeat to a lower value like 6-8 (both in AIController and sync node in the two scenes mentioned previously).
+> Another way to make this easier is to drag the defauklt sync.gd script to Sync node in both training and testing scene.
```
Comment (edbeeching) [inline docs/IMITATION_LEARNING.md:180]: typo defauklt

File: docs/IMITATION_LEARNING.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+In my case, I got:
+```Mean reward after evaluation: 5.906429767608643```
+The exact results you get may be different for various reasons, including the possibility that the hyperparameters
+and/or other variables may have changed since then.
+
+For comparison, when training just with `--rl_timesteps=250_000` I got a reward of:
+```Mean reward after evaluation: 9.194426536560059```
+
+The imitation-learned reward could be improved by tweaking hyperaparameters (the parameters provided in the script are
```
Comment (edbeeching) [inline docs/IMITATION_LEARNING.md:233]: typo hyperaparameters

File: examples/sb3_imitation.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        print("Exception while closing env: ", e)
+
+
+trajectories = []
+
+for file_path in args.demo_files:
+    with open(file_path, "r") as file:
+        data = json.load(file)
+
+    for i in range(0, len(data)):
```
Comment (edbeeching) [inline examples/sb3_imitation.py:118]: this could be just `for traj in data`, then `obs=traj[0]` and `acts=traj[1]`

no problem if you wanna leave as is though

**Comment by Ivan-267** [comment]
Thank you for the review. 

For reference and easier testing, I have pushed to a temporary branch the RL env with modifications I used in the tutorial to my fork of the examples repo: https://github.com/Ivan-267/godot_rl_agents_examples/tree/il_test_env/il_test_env (as a note switching from `SyncOverride.gd` back to `Sync.gd` would make the recording process noticeably smoother with action repeat 10). 

That one should be ready to record in the "testing scene".

**Comment by edbeeching** [comment]
Hey @Ivan-267 , I just tried it with JumperHard. The recording of trajectories works fine with the modification from the guide, but the trained model just runs backwards and forwards, I am wondering if this is because of the action space in JumperHard is a dict with three keys rather than a single vector? Do you have any ideas?

**Comment by Ivan-267** [comment]
@edbeeching Thanks for spotting this, it seems to be related to my plugin modifications not resetting done on inference and human (it seems I need to bring them back). 

In this example `done` affects the selected action, so it didn't allow the agent to e.g. jump. For training, done is reset after being sent, and for recording obs, done is also reset after the trajectory of an episode is recorded, so in those two modes it worked OK.

As for some other modifications, as recording is based on episodes, I changed the AIController reset to 1000 steps for easier recording. Here are some results of onnx inference after 3 million steps of IL training with action repeat 4 for recording / training / inference.

https://github.com/edbeeching/godot_rl_agents/assets/61947090/33367bdf-3386-4b10-bca5-256683ef5d4d

Not great yet, but working. As I finish the modifications to the plugin I'll push the update so you can try it too.

Some AIController code changes:
```gdscript
func get_action():
	return [
		float(_player.jump_action),
		_player.move_action,
		_player.turn_action
	]

func set_action(action = null):
	if action:
		_player.move_action = action["move"][0]
		_player.turn_action = action["turn"][0]
		_player.jump_action = action["jump"][0] > 0
	else:
		_player.move_action = 	clamp(
			(
				Input.get_action_strength("move_backwards")
				- Input.get_action_strength("move_forwards")
			),
			-1.0,
			0.5
		)
		_player.turn_action = (
		Input.get_action_strength("turn_left") - Input.get_action_strength("turn_right")
	)
		_player.jump_action = Input.is_action_just_pressed("jump")


func get_action_space():
	return {
		"jump": {"size": 1, "action_type": "continuous"},
		"move": {"size": 1, "action_type": "continuous"},
		"turn": {"size": 1, "action_type": "continuous"}
	}
```

Some Player.gd changes:
```gdscript
func _physics_process(_delta):
	
	if ai_controller.heuristic != "model":
		ai_controller.set_action()
```

```gdscript
func get_move_vec() -> Vector3:
	if ai_controller.done:
		return Vector3.ZERO

	return Vector3(0, 0, clamp(move_action, -1.0, 0.5))


func get_turn_vec() -> float:
	return turn_action


func get_jump_action() -> bool:
	if ai_controller.done:
		jump_action = false

	return jump_action
```

**Comment by Ivan-267** [comment]
With the updated plugin as above, 33 recorded successful episodes, action repeat 8, `n_steps=128` in PPO settings and 1 million steps of IL training:

https://github.com/edbeeching/godot_rl_agents/assets/61947090/b37bd59f-ad04-4334-8a47-5ea860142adb

**Comment by edbeeching** [comment]
Hey @Ivan-267 thanks for those comments. I managed to get it working, I think it was a combination of the issue with the done flag, not enough recorded trajectories and not enough IL training steps. 

I will review https://github.com/edbeeching/godot_rl_agents_plugin/pull/35 (tomorrow) and then we can get this merged.
